gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/AckDeadlineRenewer.java

```diff
@@ -0,0 +1,316 @@
+          && Objects.equals(other.deadline, this.deadline);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(messageId, deadline);
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("messageId", messageId)
+          .add("expectedDeadline", deadline)
+          .toString();
+    }
+  }
+
+  AckDeadlineRenewer(PubSub pubsub) {
+    PubSubOptions options = pubsub.options();
+    this.pubsub = pubsub;
+    this.executorFactory = options.executorFactory();
+    this.executor = executorFactory.get();
+    this.clock = options.clock();
+    this.messageQueue = new LinkedList<>();
+    this.messageDeadlines = new HashMap<>();
+  }
+
+  private void unsetAndScheduleNextRenewal() {
+    synchronized (futureLock) {
+      renewerFuture = null;
+      scheduleNextRenewal();
+    }
+  }
+
+  private void scheduleNextRenewal() {
+    // Schedules next renewal if there are still messages to process and no renewals scheduled that
+    // could handle them, otherwise does nothing
+    Message nextMessage;
+    synchronized (lock) {
+      Message peek = messageQueue.peek();
+      // We remove from the queue messages that were removed from the ack deadline renewer (and
+      // possibly re-added)
+      while (peek != null && (!messageDeadlines.containsKey(peek.messageId())
+          || messageDeadlines.get(peek.messageId()) > peek.expectedDeadline())) {
+        messageQueue.poll();
+        peek = messageQueue.peek();
+      }
+      nextMessage = peek;
+    }
+    synchronized (futureLock) {
+      if (renewerFuture == null && nextMessage != null) {
+        long delay =
+            (nextMessage.expectedDeadline() - clock.millis()) - NEXT_RENEWAL_THRESHOLD_MILLIS;
+        renewerFuture = executor.schedule(new Runnable() {
+          @Override
+          public void run() {
+            renewAckDeadlines();
+          }
+        }, delay, TimeUnit.MILLISECONDS);
+      }
+    }
+  }
+
+  private void renewAckDeadlines() {
+    ListMultimap<String, String> messagesToRenewNext = LinkedListMultimap.create();
+    // At every activation we renew all ack deadlines that will expire in the following
+    // RENEW_THRESHOLD_MILLIS
+    long threshold = clock.millis() + RENEW_THRESHOLD_MILLIS;
+    Message message;
+    while ((message = nextMessageToRenew(threshold)) != null) {
+      // If the expected deadline is null the message was removed and we must ignore it, otherwise
+      // we renew its ack deadline
+      if (message.expectedDeadline() != null) {
+        messagesToRenewNext.put(message.messageId().subscription(), message.messageId().ackId());
+      }
+    }
+    for (Map.Entry<String, List<String>> entry : Multimaps.asMap(messagesToRenewNext).entrySet()) {
+      // We send all ack deadline renewals for a subscription
+      pubsub.modifyAckDeadlineAsync(entry.getKey(), MIN_DEADLINE_MILLIS,
```

Comment:
I actually think that we should ask to lease a bit *longer* than the 10 seconds minimum however assume that we only got 10 seconds and use slack before we renew again.